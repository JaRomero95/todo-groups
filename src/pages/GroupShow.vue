<template>
  <div>
    <app-loading v-if="loadingGroup" />

    <div v-else>
      <div v-if="group">
        <h2
          v-if="!editing"
          class="headline"
          data-test="group-title"
        >
          {{group.name}}
          <v-icon
            @click="enableEdit"
            data-test="group-edit"
          >
            edit
          </v-icon>
        </h2>

        <group-form
          v-else
          :input-props="{appendIcon: 'clear'}"
          :input-listeners="{'click:append': disableEdit}"
          data-test="group-edit-form"
        />

        <div class="text-xs-right">
          <v-btn
            color="info"
            small
            @click="markAllTasksAsIncomplete"
            data-test="mark-all-tasks-as-incomplete"
          >
            <v-icon>check_box_outline_blank</v-icon>
            Uncheck all tasks
          </v-btn>

          <v-btn
            color="error"
            small
            @click="openDeleteDialog"
            data-test="open-delete-dialog"
          >
            <v-icon>delete</v-icon>
            Delete
          </v-btn>
        </div>

        <task-index
          v-if="group"
          :id-list="group.id"
          :tasks="tasks"
        />
      </div>

      <div v-else>
        Group not found :(
      </div>
    </div>

    <group-delete-dialog
      v-if="showDeleteDialog"
      :group="group"
      @cancel="hideDeleteDialog"
      @delete="handleDeleteGroup"
    />
  </div>
</template>

<script>
import {mapGetters, mapActions} from 'vuex';
import GroupDeleteDialog from '@/components/groups/GroupDeleteDialog.vue';
import GroupForm from '@/components/groups/GroupForm.vue';
import TaskIndex from '@/components/tasks/TaskIndex.vue';

export default {
  components: {
    GroupDeleteDialog,
    GroupForm,
    TaskIndex,
  },
  data() {
    return {
      showDeleteDialog: false,
      editing: false,
    };
  },
  computed: {
    ...mapGetters(['group', 'loadingGroup']),
    tasks() {
      return this.group && this.group.cards;
    },
  },
  created() {
    this.loadGroup(this.$route.params.id);
  },
  watch: {
    group() {
      this.disableEdit();
    }
  },
  methods: {
    ...mapActions(['loadGroup', 'deleteGroup', 'markTaskAsIncomplete']),
    async handleDeleteGroup() {
      await this.deleteGroup(this.group.id);
      this.$router.push({name: 'groups-index'});
    },
    markAllTasksAsIncomplete() {
      this.tasks.forEach(({id}) => this.markTaskAsIncomplete(id));
    },
    openDeleteDialog() {
      this.showDeleteDialog = true;
    },
    hideDeleteDialog() {
      this.showDeleteDialog = false;
    },
    enableEdit() {
      this.editing = true;
    },
    disableEdit() {
      this.editing = false;
    }
  }
}
</script>