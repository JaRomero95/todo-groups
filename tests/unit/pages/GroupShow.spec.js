import Vuex from 'vuex';
import {shallowMount, createLocalVue} from '@vue/test-utils';
import GroupShow from '@/pages/GroupShow.vue';
import stubs from '../support/stubs';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('GroupShow', () => {
  let wrapper;
  let actions;

  beforeEach(() => {
    actions = {
      loadGroup: jest.fn(),
      deleteGroup: jest.fn(),
      markTaskAsIncomplete: jest.fn(),
    };

    const store = new Vuex.Store({
      actions,
      getters: {
        group: jest.fn(() => ({
          cards: [
            {id: 11},
            {id: 12},
          ]
        })),
        loadingGroup: jest.fn(),
      },
    });

    const mocks = {
      $route: {
        params: {
          id: 19
        }
      }
    };

    wrapper = shallowMount(GroupShow, {localVue, store, stubs, mocks});
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('markAllTasksAsIncomplete', () => {
    it('invoke markTaskAsIncomplete for each task', () => {
      wrapper.find('[data-test="mark-all-tasks-as-incomplete"]').vm.$emit('click');

      expect(actions.markTaskAsIncomplete).toBeCalledTimes(2);
    });
  });
});
