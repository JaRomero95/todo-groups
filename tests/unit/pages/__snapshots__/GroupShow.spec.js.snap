// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupShow renders correctly 1`] = `
<div>
  <div>
    <div>
      <h2 data-test="group-title" class="headline">

        <v-icon-stub datatest="group-edit" data-test="group-edit">
          edit
        </v-icon-stub>
      </h2>
      <div class="text-xs-right">
        <v-btn-stub color="info" activeclass="v-btn--active" tag="button" small="true" type="button" data-test="mark-all-tasks-as-incomplete">
          <v-icon-stub>check_box_outline_blank</v-icon-stub>
          Uncheck all tasks
        </v-btn-stub>
        <v-btn-stub color="error" activeclass="v-btn--active" tag="button" small="true" type="button" data-test="open-delete-dialog">
          <v-icon-stub>delete</v-icon-stub>
          Delete
        </v-btn-stub>
      </div>
      <task-index-stub tasks="[object Object],[object Object]"></task-index-stub>
    </div>
  </div>
  <!---->
</div>
`;
